library/core/src/ptr/const_ptr.rs

```diff
@@ -258,14 +258,15 @@ impl<T: ?Sized> *const T {
         // FIXME(strict_provenance_magic): I am magic and should be a compiler intrinsic.
         //
         // In the mean-time, this operation is defined to be "as if" it was
-        // a wrapping_offset, so we can emulate it as such. This should properly
+        // two `wrapping_offset`s, so we can emulate it as such. This should properly
         // restore pointer provenance even under today's compiler.
-        let self_addr = self.addr() as isize;
-        let dest_addr = addr as isize;
-        let offset = dest_addr.wrapping_sub(self_addr);
-
-        // This is the canonical desugarring of this operation
-        self.wrapping_byte_offset(offset)
+        //
+        // This is the canonical desugaring of this operation.
+        //
+        // Note that we are using two offsets, instead of precomputing offset
+        // with `addr - self.addr()`. This is because it is easier for LLVM to
+        // optimize such code.
+        self.wrapping_byte_sub(self.addr()).wrapping_byte_add(addr)
```

Comment:
As I was playing yesterday, I noticed another possible implementation:
```suggestion
        self.mask(0).wrapping_byte_add(addr)
```
That's again just two IR instructions (<https://rust.godbolt.org/z/EeraoxKnh>), and it has the bonus of never doing a `ptrtoint`.

Any thoughts on whether that's reasonable or horrible, @nikic?